File: cf_sdk/src/main/java/com/example/cf_sdk/logic/changebanksdk/source/TransactionDatasource.java
package com.example.cf_sdk.logic.changebanksdk.source;

import com.example.cf_sdk.logic.changebanksdk.model.transaction.RecentTransactionParameter;
import com.example.cf_sdk.logic.changebanksdk.model.transaction.StatementResponse;
import com.example.cf_sdk.logic.changebanksdk.model.transaction.Transaction;
import com.example.cf_sdk.logic.changebanksdk.model.transaction.TransactionResponse;
import com.example.cf_sdk.logic.changebanksdk.parameter.transaction.PdfStatementsParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.transaction.TransactionByIdParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.transaction.TransactionsByDateParameters;
import com.example.cf_sdk.logic.changebanksdk.parameter.transaction.TransactionsSearchParameters;
import com.google.common.base.Optional;

import java.io.File;

import io.reactivex.Single;

/**
 *
 * Datasource to retrieve transactions.
 */

public interface TransactionDatasource {

    Single<TransactionResponse> getTransactionsByStatusFilter(RecentTransactionParameter transactionByStatusParameters);

    Single<TransactionResponse> getTransactionsByDate(TransactionsByDateParameters transactionsByDateParameters);

    Single<Transaction> getTransactionById(TransactionByIdParameters transactionByIdParameters);

    Single<TransactionResponse> searchTransactions(TransactionsSearchParameters transactionsSearchParameters);

    Single<Optional<File>> getPdfStatements(PdfStatementsParameters pdfStatementsParameters);

    Single<StatementResponse> getTransactionStatements(RecentTransactionParameter recentTransactionParameter);

}

File: cf_sdk/src/main/java/com/example/cf_sdk/changebankapi/source/datasource/TransactionDatasource.kt
package com.example.cf_sdk.changebankapi.source.datasource


import com.example.cf_sdk.changebankapi.model.transaction.RecentTransactionParameter
import com.example.cf_sdk.changebankapi.model.transaction.StatementResponse
import com.example.cf_sdk.changebankapi.model.transaction.Transaction
import com.example.cf_sdk.changebankapi.model.transaction.TransactionResponse
import com.example.cf_sdk.changebankapi.parameter.transaction.PdfStatementsParameters
import com.example.cf_sdk.changebankapi.parameter.transaction.TransactionByIdParameters
import com.example.cf_sdk.changebankapi.parameter.transaction.TransactionsByDateParameters
import com.example.cf_sdk.changebankapi.parameter.transaction.TransactionsSearchParameters
import com.google.common.base.Optional
import io.reactivex.Single
import java.io.File

/**
 * Created by victorojeda on 12/21/17.
 *
 *
 * Datasource to retrieve transactions.
 */
interface TransactionDatasource {
    fun getTransactionsByStatusFilter(transactionByStatusParameters: RecentTransactionParameter?): Single<TransactionResponse?>?
    fun getTransactionsByDate(transactionsByDateParameters: TransactionsByDateParameters?): Single<TransactionResponse?>?
    fun getTransactionById(transactionByIdParameters: TransactionByIdParameters?): Single<Transaction?>?
    fun searchTransactions(transactionsSearchParameters: TransactionsSearchParameters?): Single<TransactionResponse?>?
    fun getPdfStatements(pdfStatementsParameters: PdfStatementsParameters?): Single<out Optional<out File?>?>
    fun getTransactionStatements(recentTransactionParameter: RecentTransactionParameter?): Single<StatementResponse?>?
}